feat(api)-Define/Implement the detailed REST API on SettingsService for Dashboard[#253][#247][#246]

akshaylg0314 opened the pull request:
📝 PR Description
feat(api)-Define/Implement the detailed REST API on SettingsService for Dashboard[#253][#247][#246]

🔗 Related Issue
closes-> https://github.com/eclipse-pullpiri/pullpiri/issues/253,
               https://github.com/eclipse-pullpiri/pullpiri/issues/247
               https://github.com/eclipse-pullpiri/pullpiri/issues/246

🧪 Test Method
Documentation is also updated with testing methods and usage
In this PR i resolved the issue related to Clustering nodeagent also.

✅ Checklist
[✅] Code conventions are followed
[✅] Tests are added/modified
[✅] Documentation is updated (if necessary)

daeyoung-jeong-lge reviewed (changes requested):
@akshaylg0314 Please take action for my review comments.
@leeeunkoo I finished to review the codes related to REST APIs. Please review for clustering.

daeyoung-jeong-lge commented on src/server/settingsservice/borad-soc-node-rest-api.md:
```diff
@@ -0,0 +1,240 @@
+# Board/SoC/Node Management REST API Documentation
```
Thanks for the making md file for frontend developers. :-)

akshaylg0314 replied:
https://github.com/eclipse-pullpiri/pullpiri/pull/254/commits/8364dd826ef487a313613412775b915d4195c779

akshaylg0314 commented on src/server/apiserver/src/manager.rs:
```diff
@@ -4,13 +4,34 @@
-    tokio::join!(crate::route::launch_tcp_listener(), reload());
+    tokio::join!(
+        crate::route::launch_tcp_listener(),
+        start_grpc_server(),
```
This is for clustering code, grpc server in apiserver to handle req from nodeagent client.

akshaylg0314 commented on src/server/settingsservice/src/monitoring_etcd.rs:
```diff
@@ -0,0 +1,350 @@
+        .put(key, value, None)
+        .await
+        .map_err(|e| format!("ETCD put error: {}", e))?;
+
```
https://github.com/eclipse-pullpiri/pullpiri/pull/254/commits/8364dd826ef487a313613412775b915d4195c779

akshaylg0314 commented on src/server/settingsservice/src/monitoring_etcd.rs:
```diff
@@ -0,0 +1,350 @@
+use crate::monitoring_types::{BoardInfo, NodeInfo, SocInfo};
+use etcd_client::Client;
+
+type Result<T> = std::result::Result<T, String>;
```
https://github.com/eclipse-pullpiri/pullpiri/pull/254/commits/8364dd826ef487a313613412775b915d4195c779

akshaylg0314 commented on src/server/settingsservice/src/monitoring_etcd.rs:
```diff
@@ -0,0 +1,350 @@
+
+/// Get NodeInfo from etcd
+pub async fn get_node_info(node_name: &str) -> Result<NodeInfo> {
+    let mut client = get_etcd_client().await?;
```
https://github.com/eclipse-pullpiri/pullpiri/pull/254/commits/8364dd826ef487a313613412775b915d4195c779

akshaylg0314 commented:
Hey @daeyoung-jeong-lge @leeeunkoo 
I addressed all the review comments please check it once
https://github.com/eclipse-pullpiri/pullpiri/pull/254/commits/8364dd826ef487a313613412775b915d4195c779
Thanks!!

daeyoung-jeong-lge commented:
@akshaylg0314 I gave the below comment, but the latest commit doesn't include this. If this is included, please give me the location.

==================================================================================

Whenever the get_something_info(), store_something_info(), delete_something_info() functions are invoked, get_etcd_client() is also invoked. This will establish a new connection for every request, and there will be performance overhead, resource waste, and so on.

So, in my opinion, we need some approach it to be called once. Maybe we can apply something as below:

use std::sync::Arc;
use tokio::sync::OnceCell;
use etcd_client::Client;

static ETCD_CLIENT: OnceCell<Arc> = OnceCell::const_new();

pub async fn init_etcd_client() -> Result<(), MonitoringEtcdError> {
let client = Client::connect(["http://localhost:2379"], None).await?;
ETCD_CLIENT.set(Arc::new(client)).map_err(|_| MonitoringEtcdError::Other("Client already initialized".into()))?;
Ok(())
}

pub fn get_etcd_client() -> Result<Arc, MonitoringEtcdError> {
ETCD_CLIENT.get().cloned().ok_or(MonitoringEtcdError::Other("Etcd client not initialized".into()))
}

If we change like this, we need to consider how to handle connection errors or drops gracefully, and the thread safety and concurrency. For the second one, maybe we need tokio::sync::Mutex or RwLock.

akshaylg0314 commented:
> use std::sync::Arc; use tokio::sync::OnceCell; use etcd_client::Client;
> 
> static ETCD_CLIENT: OnceCell = OnceCell::const_new();
> 
> pub async fn init_etcd_client() -> Result<(), MonitoringEtcdError> { let client = Client::connect(["http://localhost:2379"], None).await?; ETCD_CLIENT.set(Arc::new(client)).map_err(|_| MonitoringEtcdError::Other("Client already initialized".into()))?; Ok(()) }
> 
> pub fn get_etcd_client() -> Result<Arc, MonitoringEtcdError> { ETCD_CLIENT.get().cloned().ok_or(MonitoringEtcdError::Other("Etcd client not initialized".into())) }

I am using etcd client from the common::etcd to manage the etcd functionality like we are doing in other components like apiserver,monitoring server ,filtergateway and so on, if you want me to do like
```
static ETCD_CLIENT: OnceCell = OnceCell::const_new();

pub async fn init_etcd_client() -> Result<(), MonitoringEtcdError> {
let client = Client::connect(["http://localhost:2379"], None).await?;
ETCD_CLIENT.set(Arc::new(client)).map_err(|_| MonitoringEtcdError::Other("Client already initialized".into()))?;
Ok(())
}

pub fn get_etcd_client() -> Result<Arc, MonitoringEtcdError> {
ETCD_CLIENT.get().cloned().ok_or(MonitoringEtcdError::Other("Etcd client not initialized".into()))
}
```
Then i will do the same.
thanks!!

daeyoung-jeong-lge reviewed (approved):
looks OK.

akshaylg0314 commented:
> Hey @leeeunkoo by this PR i also fixed the issues of missed code and build errors for clustering code now i am able to run nodeagent and apiserver with clustering functionality. Changes to do next->In nodeagent we have to implement the code for handling config file which is passed through install script right now it is taking default ip only 0.0.0.0 becuase of parsing error and file name diff: <img alt="image" width="910" height="816" src="https://private-user-images.githubusercontent.com/208792425/488783252-7bbaf927-0a6b-4d02-a17c-353c5d30c565.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTc5MTY1ODIsIm5iZiI6MTc1NzkxNjI4MiwicGF0aCI6Ii8yMDg3OTI0MjUvNDg4NzgzMjUyLTdiYmFmOTI3LTBhNmItNGQwMi1hMTdjLTM1M2M1ZDMwYzU2NS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTE1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkxNVQwNjA0NDJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kMDAxMWIwYWRiYzYxMDMxNGU4MWYyYTczODUwY2FmN2QyNTQxYmQ4ODQ4MTM1ZDY3YTRhMDRmNjA2Y2JlNmJhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.gZoiIZzOtc70JXzy38ZXc0PIGsqsBb6ZSDsAAc0aJME">
> 
> In this code of common::settings.rs it is falling back to default ip because of settings parsing error, its due to the structure we are using in install script: <img alt="image" width="558" height="549" src="https://private-user-images.githubusercontent.com/208792425/488784542-647d1ee1-9a9a-498d-89ae-1a1bf0ce5152.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTc5MTY1ODIsIm5iZiI6MTc1NzkxNjI4MiwicGF0aCI6Ii8yMDg3OTI0MjUvNDg4Nzg0NTQyLTY0N2QxZWUxLTlhOWEtNDk4ZC04OWFlLTFhMWJmMGNlNTE1Mi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTE1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkxNVQwNjA0NDJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lOWE5MjhjZDE4NTEwMjdlNDYzYWQyNTZiMTY2NTM5NTA0Y2I2NGZlZjVjOGMzNDcwZGQ4NzRlOWVkNmIzMDE0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.-vy_usTGxZpG0kTI6y8nAYJKYsyFSoH86gdQrZj9kf8">
> 
> thanks!!

Hey @leeeunkoo what should we do here? without these chnages nodeagent and clustering functionality will not work in different nodes(systems)
thanks!!
